server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java

```diff
@@ -67,13 +67,32 @@ public void getCheckpointMetadata(
         try (final GatedCloseable<SegmentInfos> segmentInfosSnapshot = indexShard.getSegmentInfosSnapshot()) {
             final Version version = segmentInfosSnapshot.get().getCommitLuceneVersion();
             final RemoteSegmentMetadata mdFile = getRemoteSegmentMetadata();
-            // During initial recovery flow, the remote store might not
-            // have metadata as primary hasn't uploaded anything yet.
-            if (mdFile == null && indexShard.state().equals(IndexShardState.STARTED) == false) {
-                listener.onResponse(new CheckpointInfoResponse(checkpoint, Collections.emptyMap(), null));
-                return;
+
+            // Handle null metadata file case
+            if (mdFile == null) {
+                // During initial recovery flow, the remote store might not
+                // have metadata as primary hasn't uploaded anything yet.
+                if (indexShard.state().equals(IndexShardState.STARTED) == false) {
+                    // Non-started shard during recovery
+                    listener.onResponse(new CheckpointInfoResponse(checkpoint, Collections.emptyMap(), null));
+                    return;
+                } else if (indexShard.routingEntry().isSearchOnly()) {
+                    // Allow search-only replicas to become active without metadata
+                    logger.debug("Search-only replica proceeding without remote metadata: {}", indexShard.shardId());
+                    listener.onResponse(
+                        new CheckpointInfoResponse(indexShard.getLatestReplicationCheckpoint(), Collections.emptyMap(), null)
+                    );
+                    return;
+                } else {
+                    // Regular replicas should not be active without metadata
+                    listener.onFailure(
+                        new IllegalStateException("Remote metadata file can't be null for active regular replica: " + indexShard.shardId())
```

Comment:
I'm ok changing the assert to a failure case here as in prod this would be an NPE, but lets preserve the message.
```suggestion
                        new IllegalStateException("Remote metadata file can't be null if shard is active: " + indexShard.shardId())
```